blog/2021-06-29-transaction-presigning.md

```diff
@@ -0,0 +1,148 @@
+
+If we want to allow users to sell XMR for BTC, then we need a protocol where party A locks up XMR first. 
+We [developed such a protocol](https://arxiv.org/abs/2101.12332) but in order for it to be secure from a game-theoretic PoV, we need to pre-sign spending transaction from a Monero joint output.
+
+In a nutshell, if we want to provide a non-custodial trading solution where users can buy AND sell XMR for BTC, we need transaction pre-signing.
+
+## It doesn't work on present day Monero
+
+Unfortunately, it is not possible to efficiently implement pre-signing of transactions in present day Monero. 
+The reason for that is very technical and we believe it can be fixed but likely requires a hard-fork.
+
+Lucas recently [made an effort](/blog/2021/05/19/monero-transaction) in documenting the transaction structure and signature algorithms of Monero.
+I'll try to reiterate the necessary concepts here but if you want more information on how things work, I suggest to take a look at Lucas' post.
+
+To sign a transaction in Monero, one needs to choose ten other outputs that - together with the output you want to spend - form the ring for Monero's CLSAG algorithm.
+These outputs are represented as what is called *key-offsets* for efficiency reasons.
+For the ring itself, we need the public key of each output but public keys are relatively big in size.
+32 bytes even in their compressed form to be precise.
+Times 11, that would be 352 bytes *per input* just to represent the ring for the signature.
+
+To be more space efficient, we instead enumerate all outputs in the blockchain and create a list of offsets where the first element in the list is an absolute output index and all subsequent elements are relative to their previous index.
+This is what Monero calls *key-offsets*.
+Storing the *key-offsets* within the transaction is much more space efficient because a) the values are much smaller and b) we can store them as variable length integers.
+
+The problem is that this list of key-offsets is also incorporated into the transaction's signature hash.
+
+Why is this a problem?
+Imagine two transactions, `A` and `B`:
+
+- `B` spends an output from `A` and we would like to pre-sign this spending transaction.
+- To sign `B`, we need to compute the signature hash and run the CLSAG algorithm using this hash.
+- To compute the signature hash, we need to come up with the key-offsets of our ring.
+- To compute the key-offsets for the ring, we need to know, what the output index of `A` is.
+- To know the output index of `A`, the transaction needs to be picked up by a miner and included in the blockchain.
+
+Remember that the whole point of pre-signing `B` is that we have a transaction with valid signatures **prior** to the point of signing and broadcasting `A`.
+Because the key-offsets are included in the signature hash, we can't produce it up until the output we want to spend is included in the blockchain.
```

Comment:
To close this section you could add one more sentence to repeat what you just said: 
e.g. 
```

Differently said: we cannot pre-sign transaction `B` which spends from transaction `A` before transaction `A` has been included in the blockchain.
```